File: chapters/09-api-design/go/cursor_pagination.go
// Scenario: GET /orders?cursor=... for high-volume list API (no OFFSET on huge tables)
// Demonstrates: opaque cursor encoding keyset pagination
// Trade-off: cursor invalidation on sort change — version cursor schema in API contract

package main

import (
	"encoding/base64"
	"strings"
)

type OrderRow struct {
	ID        string
	CreatedAt string
}

type Page[T any] struct {
	Items      []T
	NextCursor string
}

func encodeCursor(orderID, createdAt string) string {
	raw := orderID + "|" + createdAt
	return base64.RawURLEncoding.EncodeToString([]byte(raw))
}

func decodeOrderID(cursor string) string {
	b, _ := base64.RawURLEncoding.DecodeString(cursor)
	parts := strings.SplitN(string(b), "|", 2)
	return parts[0]
}

func pageOrders(fetched []OrderRow, limit int) Page[OrderRow] {
	hasMore := len(fetched) > limit
	items := fetched
	if hasMore {
		items = fetched[:limit]
	}
	var next string
	if hasMore {
		last := items[len(items)-1]
		next = encodeCursor(last.ID, last.CreatedAt)
	}
	return Page[OrderRow]{Items: items, NextCursor: next}
}

File: docs/chapters/09-api-design/java/CursorPagination.java
// Scenario: GET /orders?cursor=... for high-volume list API (no OFFSET on huge tables)
// Demonstrates: opaque cursor encoding keyset pagination
// Trade-off: stable under inserts vs OFFSET simplicity for admin-only small pages

import java.nio.charset.StandardCharsets;
import java.util.Base64;
import java.util.List;

record OrderRow(String id, String createdAt) {}
record Page<T>(List<T> items, String nextCursor) {}

final class CursorPagination {
    static String encodeCursor(String orderId, String createdAt) {
        String raw = orderId + "|" + createdAt;
        return Base64.getUrlEncoder().withoutPadding()
            .encodeToString(raw.getBytes(StandardCharsets.UTF_8));
    }

    static String decodeOrderId(String cursor) {
        String raw = new String(Base64.getUrlDecoder().decode(cursor), StandardCharsets.UTF_8);
        return raw.split("\\|", 2)[0];
    }

    // Repository would: WHERE (created_at, id) < (?, ?) ORDER BY created_at DESC, id DESC LIMIT n+1
    static Page<OrderRow> page(List<OrderRow> fetched, int limit) {
        boolean hasMore = fetched.size() > limit;
        List<OrderRow> items = hasMore ? fetched.subList(0, limit) : fetched;
        String next = hasMore
            ? encodeCursor(items.get(items.size() - 1).id(), items.get(items.size() - 1).createdAt())
            : null;
        return new Page<>(items, next);
    }
}
